File: crossplane-dapr/producer-springboot/src/main/java/com/agnostic/producerspringboot/PublishOrderRequest.java
package com.agnostic.producerspringboot;

import jakarta.validation.constraints.DecimalMin;
import jakarta.validation.constraints.NotBlank;
import java.math.BigDecimal;

record PublishOrderRequest(
    @NotBlank(message = "id must not be blank")
    String id,
    @DecimalMin(value = "0.01", inclusive = true, message = "amount must be greater than zero")
    BigDecimal amount
) {}

File: crossplane-dapr/producer/src/main/kotlin/com/agnostic/producer/dto/PublishOrderRequest.kt
package com.agnostic.producer.dto

import com.agnostic.common.events.OrderCreatedV1
import com.agnostic.common.serialization.BigDecimalAsNumberSerializer
import kotlinx.serialization.Serializable
import java.math.BigDecimal

@Serializable
data class PublishOrderRequest(
    val id: String,
    @Serializable(with = BigDecimalAsNumberSerializer::class)
    val amount: BigDecimal,
) {
    fun validate() {
        require(id.isNotBlank()) { "id must not be blank" }
        require(amount > BigDecimal.ZERO) { "amount must be greater than 0" }
    }

    fun toEvent(): OrderCreatedV1 = OrderCreatedV1(id = id, amount = amount)
}
